### PR TITLE
OVDB-136: Point Group Inspection

### PR DESCRIPTION
**Pull request by danrbailey**
Move group inspection functionality from point_group_internal::GroupInfo up into the AttributeSet and Descriptor so that it becomes part of the public API and can be used outside of the point group methods.

### CODE REVIEW

File: openvdb/points/AttributeSet.h [hunk 1]
```diff
@@ -375,7 +381,8 @@ class OPENVDB_API AttributeSet::Descriptor
     /// Return @c true if group exists
     bool hasGroup(const Name& group) const;
     /// Define a group name to offset mapping
-    void setGroup(const Name& group, const size_t offset);
+    void setGroup(const Name& group, const size_t offset,
+        const bool checkValidOffset = false);
```
Comment (Idclip) [inline openvdb/points/AttributeSet.h]: `@param` for these arguments

File: openvdb/points/AttributeSet.h [hunk 2]
```diff
@@ -402,6 +409,27 @@ class OPENVDB_API AttributeSet::Descriptor
...
+
+    /// Return @c true if there are sufficient empty slots to allow compacting
+    bool canCompactGroups() const;
+
+    /// Return the next empty group slot
+    size_t nextUnusedGroupOffset() const;
+
+    /// Determine if a move is required to efficiently compact the data and store the
+    /// source name, offset and the target offset in the input parameters
+    bool requiresGroupMove(Name& sourceName, size_t& sourceOffset, size_t& targetOffset) const;
```
Comment (Idclip) [inline openvdb/points/AttributeSet.h]: same here

File: openvdb/points/PointGroup.h [hunk 3]
```diff
@@ -641,14 +521,14 @@ inline void compactGroups(PointDataTree& tree)
         LeafManagerT leafManager(tree);
         tbb::parallel_for(leafManager.leafRange(), copy);
 
-        descriptor->setGroup(sourceName, targetOffset);
+        descriptor->setGroup(sourceName, targetOffset, /*checkValidOffset=*/true);
```
Comment (Idclip) [inline openvdb/points/PointGroup.h]: Why does `checkValidOffset` need to be true here - doesn't this algorithm guarantee this?

File: openvdb/points/PointGroup.h [hunk 4]
```diff
@@ -641,14 +521,14 @@ inline void compactGroups(PointDataTree& tree)
...
     }
 
     // drop unused attribute arrays
 
-    std::vector<size_t> indices = groupInfo.populateGroupIndices();
+    std::vector<size_t> indices = attributeSet.groupAttributeIndices();
```
Comment (Idclip) [inline openvdb/points/PointGroup.h]: can be const

**Comment by Idclip** [comment]
Minor comments, main one about the use of `checkValidOffset` otherwise lgtm

File: openvdb/points/PointGroup.h (hunk 3 above)
Comment (danrbailey) [reply openvdb/points/PointGroup.h]: I believe it to be just an abundance of caution. I've addressed all feedback now.

**Comment by danrbailey** [comment]
Thanks a lot for reviewing @Idclip!

**Pull request merged**